lms/djangoapps/grades/signals.py

```diff
@@ -122,6 +122,31 @@ def recalculate_subsection_grade_handler(sender, **kwargs):  # pylint: disable=u
        - course_id: Unicode string representing the course
        - usage_id: Unicode string indicating the courseware instance
     """
+    def ensure_subsection(usage_id):
+        """
+        Ensures that the we update the persistent *subsection* score for a
+        given signal, even if the signal was sent by a block "lower" on the
+        course structure tree.
+        """
+        # TODO: efischer WIP - there has to be a better way!
+        usage_key = UsageKey.from_string(usage_id).replace(course_key=course_key)
+        if usage_key.block_type == 'sequential':
+            return usage_key
+
+        from openedx.core.djangoapps.content.block_structure.api import get_block_structure_manager
+        manager = get_block_structure_manager(course_key)
+        block_structure = manager.get_collected()
+
+        parents = block_structure.get_parents(usage_key)
+        while usage_key.block_type != 'sequential' and len(parents) > 0:
+            usage_key = parents[0]  # Note - this is naively/incorrectly assuming get_parents returns a list of length 1
```

Comment:
Answer: depth first, as that will likely be fastest in the "usual" case. Here are the assumptions being made:

- treat the input block as the root of a DFS tree
- the tree is navigated by following `get_parent` links
- DFS tree has leaf nodes where `len(get_parents)` is 0, or `block_type` is `sequential`. Note that if a subsection contains another subsection, only the "lower" one will be updated when scores under it change.
- The full DFS tree will be exhausted, and all `sequential` blocks found will have their persistent scores updated.